### PR TITLE
Teet 889

### PR DESCRIPTION
**Pull request by jkorpelin**
Admin view and functionality for managing users

### CODE REVIEW

File: app/backend/src/clj/teet/admin/admin_queries.clj [hunk 1]
```diff
@@ -6,21 +6,103 @@
...
+                       (and
+                         [?lc :thk.lifecycle/activities ?act]
+                         [?act :activity/tasks ?task]
+                         [?task :task/assignee ?u])
+                       [?project :thk.project/owner ?u]
+                       (and
+                         [?lc :thk.lifecycle/activities ?act]
+                         [?act :activity/manager ?u]))]}))
+
+(defn user-list-query
```
Comment (erno) [inline app/backend/src/clj/teet/admin/admin_queries.clj:73]: This search-clause multimethod / reduce / search map construction could benefit from an outlining comment.

File: app/backend/src/clj/teet/admin/admin_commands.clj [hunk 2]
```diff
@@ -54,38 +55,81 @@
...
-             permission-map))))
-
 (defcommand :admin/create-user
   {:doc "Create user (although for now can be also used to add global permissions to existing users through admin view)"
    :context {:keys [conn user db]}
    :payload user-data
    :project-id nil
    :pre [(:user/person-id user-data)
-         (user-spec/estonian-person-id? (:user/person-id user-data))]
+         (user-spec/estonian-person-id? (:user/person-id user-data))
```
Comment (outamaa) [inline app/backend/src/clj/teet/admin/admin_commands.clj:64]: Could this be specified as part of the command payload spec?

File: app/backend/src/clj/teet/admin/admin_commands.clj [hunk 3]
```diff
@@ -54,38 +55,81 @@
...
-                    {})]
-                 ;; New user, no need to check
-                 [(merge (user-model/new-user user-person-id)
-                         (when global-permission
-                           (new-permission user
-                                           global-permission
-                                           (java.util.Date.))))]))})
+                   global-role (:user/global-role user-data)]
+               ;; New user, no need to check
+               (println "tx: " (pr-str [(merge (user-model/new-user user-person-id)
```
Comment (outamaa) [inline app/backend/src/clj/teet/admin/admin_commands.clj:72]: Should be removed or changed to proper log fn.

File: app/backend/src/clj/teet/admin/admin_queries.clj [hunk 4]
```diff
@@ -6,21 +6,103 @@
             [teet.environment :as environment]
             [clojure.walk :as walk]))
 
+(defmulti search-clause (fn [[field _value]] field))
```
Comment (outamaa) [inline app/backend/src/clj/teet/admin/admin_queries.clj:9]: Nice use of multimethods.

File: app/backend/src/clj/teet/admin/admin_queries.clj (hunk 1 above)
Comment (outamaa) [reply app/backend/src/clj/teet/admin/admin_queries.clj:73]: Yeah this is pretty heavy to parse.

File: app/backend/src/clj/teet/migration/user_permissions.clj [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+;; sort by role importance and take first
+;; create transaction with call to (set-global-role) for all users with their hightest permission role
+
+(def permission-importance
+  {:admin 1
+   :manager 2
+   :internal-consultant 3
+   :external-consultant 4})
+
+(defn remove-overlapping-global-rights
```
Comment (outamaa) [inline app/backend/src/clj/teet/migration/user_permissions.clj:17]: Here as well, very hard to parse the implementation, and not sure what's meant by overlapping global rights. So docstring and maybe a few helpful comments in the threading macro.

**Review by outamaa — APPROVED** [review]
Got about halfway through the code at this point. :)
